### PR TITLE
Mat 329/330

### PR DESCRIPTION
**Pull request by stuartarchibald**
This PR adds patches for a couple of issues:
- Removes all caching of the JNI environment pointer. Even though the actual execution using the pointer was safe via the ordering in which it was performed, it was rather confusing. Removing the cache and always obtaining a local JNIEnv pointer regardless of context means the code is much clearer at the cost of a tiny loss in performance.
- Disables tests that rely on the impairing of `AttachCurrentThread()` to break the `getEnv()` mechanism, all `JNIEnv` calls now go through the `getEnv()` function now, including those which are statically initialised. Therefore causing impairment to the `AttachCurrentThread()` function breaks the initialisers before the code to be tested can be run.
- Deletes a local ref that escaped. There's no harm in leaving it but might as well keep things clean if possible.

Coverage:
Java: 79% (no change)
build/src/jshim 0.8% (no change)
build/src/librdag 16.1%  (no change)
src/jshim 57.6% (down 0.3%, due to method removal)
src/librdag 91.8% (no change)
src/librdag/runners 79% (no change).

### CODE REVIEW

File: src/jshim/jvmmanager.cc [hunk 1]
```diff
@@ -62,8 +62,10 @@ JVMManager::initialize(JavaVM *jvm)
 {
   VAL64BIT_PRINT("VM pointer", jvm);
 
+  JNIEnv * env;
+
   // Attempt to get env for JNI version 1.2
-  int ret = jvm->GetEnv((void **)&_env, JNI_VERSION_1_2);
+  int ret = jvm->GetEnv((void **)&env, JNI_VERSION_1_2);
```
Comment (gmarkall) [inline src/jshim/jvmmanager.cc:68]: The env that we're getting here looks unused - do we need this call to remain?

Comment (gmarkall) [inline src/jshim/jvmmanager.cc:68]: Summary of out-of-band conversation - we're leaving it here to check the JNI environment has the correct version available.

**Comment by gmarkall** [comment]
This looks good. After some discussion:
- We think we can eliminate some calls to `JVMManager::GetEnv()` by passing the environment obtained with `jvm->GetEnv()` in `JVMManager::initialize()` to the `registerReferences()` function.
- This has the advantage that we're reducing the number of calls to `JVMManager::getEnv()` with the goal of eventually removing it in favour of passing the JIN environment pointer provided to us by the JVM in the JNI call, which should generally be safer.
- However, as there are a couple of use cases for `JVMManager::getEnv()` that are a little more fiddly to remove, and there may be a future need for this function, we're going to leave it in and remove at a later stage if no more use-cases are found.
- Generally, we don't expect to need to get an environment at points in the code where we don't already have one - the typical case where we may need one is for using for reporting an error back to the JVM, but we have a mechanism for doing this relatively cleanly with C++ exceptions and the C++->Java exception translation logic.

**Comment by stuartarchibald** [comment]
Thanks for reviewing. In response to your comments, I have now refactored the offending code such that the `JVMManager::getEnv()` function is unused. In places where a `JNIEnv *` is needed the reference is obtained via passing the pointer through the function args. A review of the usefulness of the `JVMManager::getEnv()` function is slated for 6 month's time, ticket [[MAT-406]](http://jira.opengamma.com/browse/MAT-406).

Coverage of `src/jshim` drops 0.1% as a result of refactor.

**Comment by gmarkall** [comment]
This looks great! Happy to merge.

**Pull request merged**